### PR TITLE
runwasi: Use youki to fully support OCI runtime spec in wasmedge

### PR DESCRIPTION
**Pull request by utam0k**
Partially Fix: https://github.com/containerd/runwasi/issues/23 (After this PR, I'm planning to create a PR to address wasmtime.)

Remaining issues
- [x] Wait for merge:
    - [x] https://github.com/containers/youki/pull/1637
    - [x] https://github.com/containers/youki/pull/1403
- [x] Pass all tests
- [x] Pass CIs

### CODE REVIEW

File: .gitmodules
Comment (utam0k) [inline .gitmodules:1]: It is for the efficiency of development and debugging and will eventually be deleted.

**Comment by utam0k** [comment]
Current Status.
```console
$ cargo test --no-run --message-format=json | jq -r "select(.profile.test == true) | .filenames[]" | while read -r bin; do;
 sudo LD_LIBRARY_PATH=/home/utam0k/.wasmedge/lib "$bin"
done
[...]
running 3 tests
test instance::tests::test_maybe_open_stdio ... ok
test instance::wasitest::test_delete_after_create ... ok
test instance::wasitest::test_wasi ... ok

test result: ok. 3 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.14s
```

File: Dockerfile
Comment (ipuustin) [inline Dockerfile:19]: I think libcontainer doesn't depend on libsystemd-dev or libdbus-glib-1-dev? Or does it?

File: crates/wasmedge/src/instance.rs
Comment (ipuustin) [inline crates/wasmedge/src/instance.rs:200]: Are the oci functions such as `get_cgroup`, `setup_cgroup` etc. used elsewhere, or can we drop them too?

**Review by ipuustin — COMMENTED** [review]
Looks cool! Couple of questions:
1. You say this partially fixes #23 -- can you tell which parts will be missing?
2. Does this cause any functional changes outside of the spec? I mean, does Youki use the three-process model now for running the VM? Is the namespace isolation the same?

I was preparing some seccomp support in https://github.com/ipuustin/runwasi/commits/seccomp. I'm thinking I'll probably pause that work until this PR merges and then add the unit tests and other potentially useful stuff on top of this PR.

File: Cargo.toml
Comment (utam0k) [reply Cargo.toml:32]: Thanks for your comment. I'll check it later

File: Dockerfile
Comment (utam0k) [reply Dockerfile:19]: `libcontainer` support systemd-cgroup driver, so it depends on them.

File: crates/wasmedge/src/instance.rs
Comment (utam0k) [reply crates/wasmedge/src/instance.rs:200]: Since `libcontainer` supports set-up cgroup, I think don't need to take care of cgroup by runwasi.

**Comment by utam0k** [comment]
:wave: Hi, @ipuustin. Thanks for your comment.

We need to support wasmtime after this PR.
> You say this partially fixes https://github.com/containerd/runwasi/issues/23 -- can you tell which parts will be missing?

Basically the same. Youki implements the oci-spec following runc's implementation, but more simply.
> Does this cause any functional changes outside of the spec? I mean, does Youki use the three-process model now for running the VM? Is the namespace isolation the same?

I agree with you. Sorry for the inconvenience.
> I was preparing some seccomp support in https://github.com/ipuustin/runwasi/commits/seccomp. I'm thinking I'll probably pause that work until this PR merges and then add the unit tests and other potentially useful stuff on top of this PR.

File: Cargo.toml
Comment (utam0k) [reply Cargo.toml]: It looks youki that was on this PR was too old.

File: .github/workflows/ci.yml
Comment (utam0k) [inline .github/workflows/ci.yml:24]: See: 
https://github.com/containerd/runwasi/pull/78/files#r1118520073

File: Dockerfile
Comment (ipuustin) [reply Dockerfile:19]: Ok I see, there's the `systemd` feature for libcontainer.

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (ipuustin) [inline crates/containerd-shim-wasmedge/src/instance.rs:352]: Just drop commented out code -- git will remember those if need be.

**Review by ipuustin — CHANGES_REQUESTED** [review]
I think this concept looks good! When do you think next libcontainer release will be -- should we wait until that or use the libcontainer commit from main branch?

File: .github/workflows/ci.yml
Comment (utam0k) [reply .github/workflows/ci.yml:43]: You need to run it as a root user.

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (utam0k) [reply crates/containerd-shim-wasmedge/src/instance.rs:212]: If a container is present to meet OCI specifications, it must return Ok.

File: .github/workflows/ci.yml
Comment (ipuustin) [reply .github/workflows/ci.yml:43]: Yes I see! You could maybe do something like this to run the tests with sudo:

```patch
diff --git a/crates/containerd-shim-wasm/src/sandbox/mod.rs b/crates/containerd-shim-wasm/src/sandbox/mod.rs
index e807480..72d7ac2 100644
--- a/crates/containerd-shim-wasm/src/sandbox/mod.rs
+++ b/crates/containerd-shim-wasm/src/sandbox/mod.rs
@@ -16,5 +16,4 @@ pub use shim::{Cli as ShimCli, Local};

 pub mod oci;

-#[cfg(test)]
-mod testutil;
+pub mod testutil;
diff --git a/crates/containerd-shim-wasm/src/sandbox/testutil.rs b/crates/containerd-shim-wasm/src/sandbox/testutil.rs
index a2c1f86..baa61ee 100644
--- a/crates/containerd-shim-wasm/src/sandbox/testutil.rs
+++ b/crates/containerd-shim-wasm/src/sandbox/testutil.rs
@@ -41,6 +41,7 @@ pub fn run_test_with_sudo(test: &str) -> Result<()> {
         .map_err(Error::from)
 }

+#[macro_export]
 macro_rules! function {
     () => {{
         fn f() {}
@@ -53,4 +54,4 @@ macro_rules! function {
     }};
 }

-pub(crate) use function;
+pub use function;
diff --git a/crates/containerd-shim-wasmedge/src/instance.rs b/crates/containerd-shim-wasmedge/src/instance.rs
index 64c9987..7e1b8f5 100644
--- a/crates/containerd-shim-wasmedge/src/instance.rs
+++ b/crates/containerd-shim-wasmedge/src/instance.rs
@@ -256,6 +256,9 @@ mod wasitest {
     use std::sync::mpsc::channel;
     use std::time::Duration;

+    use containerd_shim_wasm::function;
+    use containerd_shim_wasm::sandbox::exec::has_cap_sys_admin;
+    use containerd_shim_wasm::sandbox::testutil::run_test_with_sudo;
     use oci_spec::runtime::{ProcessBuilder, RootBuilder, SpecBuilder};
     use tempfile::{tempdir, TempDir};

@@ -383,6 +386,10 @@ mod wasitest {
     #[test]
     #[serial]
     fn test_wasi() -> Result<(), Error> {
+        if !has_cap_sys_admin() {
+            println!("running test with sudo: {}", function!());
+            return run_test_with_sudo(function!());
+        }
         let dir = tempdir()?;
         let path = dir.path();
         let wasmbytes = wat2wasm(WASI_HELLO_WAT).unwrap();
@@ -401,6 +408,10 @@ mod wasitest {
     #[test]
     #[serial]
     fn test_wasi_error() -> Result<(), Error> {
+        if !has_cap_sys_admin() {
+            println!("running test with sudo: {}", function!());
+            return run_test_with_sudo(function!());
+        }
         let dir = tempdir()?;
         let wasmbytes = wat2wasm(WASI_RETURN_ERROR).unwrap();
```

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (ipuustin) [reply crates/containerd-shim-wasmedge/src/instance.rs:212]: Sure, I was just thinking that they are called one after another, so this could maybe be a place for optimization. But this is a very minor thing and I'm fine either way.

**Comment by utam0k** [comment]
> I think this concept looks good! When do you think next libcontainer release will be -- should we wait until that or use the libcontainer commit from main branch?

@ipuustin Thanks for your comment. We may release it soon. However, I would like to merge once with this commit as specified. This is because I am afraid of conflicts. I have already done several major conflict resolutions on this PR.

File: .github/workflows/ci.yml
Comment (utam0k) [reply .github/workflows/ci.yml:43]: @ipuustin I love this idea. I have applied for it.

File: crates/containerd-shim-wasmedge/src/executor.rs
Comment (ipuustin) [inline crates/containerd-shim-wasmedge/src/executor.rs:29]: Hmm, what if `args` is empty?

File: crates/containerd-shim-wasmedge/src/executor.rs
Comment (ipuustin) [inline crates/containerd-shim-wasmedge/src/executor.rs:86]: This is an interesting function especially if we will later on do things such as map workloads to runtimes which can support the features the workload uses.

**Review by ipuustin — APPROVED** [review]
I think this looks good now. Going forward we need to add a bunch of tests trying out the various OCI spec features, which I suppose will be good for both this project and libcontainer too.

File: crates/containerd-shim-wasmedge/src/executor.rs
Comment (utam0k) [reply crates/containerd-shim-wasmedge/src/executor.rs:29]: Maybe we need to validate it?

File: crates/containerd-shim-wasmedge/src/executor.rs
Comment (utam0k) [reply crates/containerd-shim-wasmedge/src/executor.rs:86]: Initially, this trait is designed for youki.
https://github.com/containers/youki/blob/2260c1ea776584140c659d166888addaa73b936d/crates/youki/src/workload/wasmedge.rs#L51-L63

File: README.md
Comment (devigned) [inline README.md:1]: Thank you so much for adding to the comments. The additional context is super helpful for a naive observer approaching the topic.

**Review by devigned — CHANGES_REQUESTED** [review]
This PR looks awesome. Thank you so much, @utam0k!!

Before it is merged, I think we would like to resolve: https://github.com/containerd/runwasi/pull/78#discussion_r1140069567, and I'd like @cpuguy83 to give it a review.

Really fantastic work here. Thank you again.

**Comment by utam0k** [comment]
@devigned Thanks for your review. PTAL

File: crates/containerd-shim-wasm/src/sandbox/shim.rs
Comment (cpuguy83) [inline crates/containerd-shim-wasm/src/sandbox/shim.rs:42]: I can't help but think this should be specified by the implementation, ie `/run/io.containerd.wasmedge.v1` OR from the spec?

Comment (cpuguy83) [reply crates/containerd-shim-wasm/src/sandbox/shim.rs:42]: It's also not clear to me what this is used for

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Review by devigned — APPROVED** [review]
lgtm!

I'm +1 pending @cpuguy83 feedback. Great work, @utam0k!

File: crates/containerd-shim-wasm/src/sandbox/shim.rs
Comment (utam0k) [reply crates/containerd-shim-wasm/src/sandbox/shim.rs:42]: This directory is the same role as runc; it stores files for managing the state of the container and signaling start.
If runwasi is to be considered only as an OCI runtime, I don't think the name containerd should be included in the directory. I think that including it may compromise its versatility.
https://github.com/opencontainers/runc/blob/efad7a3b80f4585065f93aad3e91989da5012d63/man/runc.8.md?plain=1#L108-L111

Comment (cpuguy83) [reply crates/containerd-shim-wasm/src/sandbox/shim.rs:42]: In our case containerd is passing down a state dir to the shim that we should be using rather than coming up with our own.

I can't exactly recall how we setting that dir right now but since this crate is only a library I think we should leave any sort of state dir to a higher level.

**Comment by cpuguy83** [comment]
Requesting t hat we keep all other merges on hold until this is in.
Looks like we need to regenerate Cargo.lock due to a merge conflict.

File: crates/containerd-shim-wasm/src/sandbox/shim.rs
Comment (utam0k) [reply crates/containerd-shim-wasm/src/sandbox/shim.rs:42]: @cpuguy83 Thanks for your comment 🙇 That is sort of an ideal, but when it comes to adjusting to reality it takes time. Many OCI Runtimes already require a state command. This command is inevitably stateful. Also, it is an OCI specification that `create` command and `start` command are two different commands: `create` creates the container process and the `start` command sent a message to that process to start the container. In other words, the `start` command must leave an interface somewhere to interact with the container created by create. That interface is left in this directory.
https://containers.github.io/youki/user/liboci_cli.html

**Comment by utam0k** [comment]
> Requesting t hat we keep all other merges on hold until this is in. Looks like we need to regenerate Cargo.lock due to a merge conflict.

Sorry for a lof of changes. I have resolved it.

File: crates/containerd-shim-wasm/src/sandbox/shim.rs
Comment (cpuguy83) [reply crates/containerd-shim-wasm/src/sandbox/shim.rs:42]: @utam0k Yes I understand that.
For us, containerd passes down the state.
As an example when you create a container with containerd, it's not using the default runc state dir because containerd (or more precisely whoever started the container) is managing that path.

If it's docker `/run/docker/runtime-runc/moby` (where `moby` is the containerd namespace) because docker customizes the path when it invokes containerd. 

```console
# jq . /run/containerd/io.containerd.runtime.v2.task/moby/01e48c19c2af1da405f209b79fcc10542372ab82fb081ad80f713826797e6ded/options.json
{
  "binary_name": "runc",
  "root": "/var/run/docker/runtime-runc",
  "systemd_cgroup": true
}
```

If it's ctr it's `/run/containerd/runc/default/test` (where `default` is the containerd namespace and `test` is the container id).
ctr does not customize this path so it's just the default.

```console
# jq . /run/containerd/io.containerd.runtime.v2.task/default/test/options.json 
{}
```

So we should expect the actual shim implementation to pass down this state dir from higher up the stack.
The above mentioned directories is the OCI bundle path.
So we should be able to check for a state dir in this file and 

We can choose a default path, but it also must be namespaced by the containerd namespace or else we can end up with conflicts.

Comment (cpuguy83) [reply crates/containerd-shim-wasm/src/sandbox/shim.rs:42]: The other part of this is we have a long running daemon managing state and should not need to persist it across calls.

When the daemon dies so do the containers it is managing.

Comment (utam0k) [reply crates/containerd-shim-wasm/src/sandbox/shim.rs]: @cpuguy83 Make sense a lot 💯 I have updated my PR!

**Comment by utam0k** [comment]
@cpuguy83 PTAL

**Review by Mossaka — COMMENTED** [review]
Mostly LGTM! I have some questions and most of them are just for me to understand `libcontainer` more. I would greatly appreicate your answers.

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (cpuguy83) [inline crates/containerd-shim-wasmedge/src/instance.rs:142]: We also need to account for the containerd namespace otherwise we could end up with 2 containers with the same ID but different namespaces writing to the same state dir.

**Review by cpuguy83 — CHANGES_REQUESTED** [review]
I think as a whole I was hoping to see this integration at a lower level than youki's Executor interface so that we can have more control over the runtime environment and reduce on-disk state since we aren't managing linux processes (outside of what's needed to deal with cgroups).
I think for now (after the requested changes are made) this will work out well.

It will be interesting to see projects like slight and spin adopt similar changes since they are more function handler based and spinning up a new process per function is not what would be desired.
But again, as is this is probably fine and we can always adjust as needed.

**Comment by cpuguy83** [comment]
@rumpl

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (utam0k) [reply crates/containerd-shim-wasmedge/src/instance.rs:212]: It can be done; youki actually does. 
https://github.com/containers/youki/blob/d1e19fdbe3229999c287949e31615b4251c90d17/crates/youki/src/workload/executor.rs#L3-L13

At an abstract level, you can implement what is to be executed in the container.

File: Cargo.toml
Comment (utam0k) [reply Cargo.toml:34]: Yes, it did but we haven't released the latest version runwasi needs yet. 
https://crates.io/crates/libcontainer

File: .github/workflows/ci.yml
Comment (utam0k) [reply .github/workflows/ci.yml:26]: Make sense

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (utam0k) [reply crates/containerd-shim-wasmedge/src/instance.rs:230]: @Mossaka Sorry, may I ask you to elaborate a bit more on your question?

**Comment by utam0k** [comment]
> I think as a whole I was hoping to see this integration at a lower level than youki's Executor interface so that we can have more control over the runtime environment and reduce on-disk state since we aren't managing linux processes (outside of what's needed to deal with cgroups). I think for now (after the requested changes are made) this will work out well.
> 
> It will be interesting to see projects like slight and spin adopt similar changes since they are more function handler based and spinning up a new process per function is not what would be desired. But again, as is this is probably fine and we can always adjust as needed.

@cpuguy83 
I don't know much about `slight` and `spin`, so I can't comment much.
Just, the biggest weakness of this architecture is its strong dependence on `youki`/`libcontainer`. There are many ways to think about this. The license is quite loose, so if something goes wrong, you can fork in the worst case. I can also participate in the development of runwasi if needed, but only in my spare time (youki is my hobby too, and being vendor neutral is what I would like). I look forward to a future with runwasi.

**Comment by utam0k** [comment]
@Mossaka @cpuguy83 Thanks for your review. I know it is hard to review because the changes are so significant. PTAL

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (utam0k) [reply crates/containerd-shim-wasmedge/src/instance.rs]: I made it consider the namespace.
https://github.com/containerd/runwasi/pull/78/commits/3fc44d2df7a0e3d52a174d92e6316909b87c7c62

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (Mossaka) [reply crates/containerd-shim-wasmedge/src/instance.rs:225]: Thanks for the link! It looks like three out of 4 executors are wasm executors. I am assuming the default executor is the one that can execute normal container workloads, right?

**Comment by rumpl** [comment]
I didn't have the time to take a look at the code yet, will look closely tomorrow but one thing that comes to mind is: why are we using this at the wasmedge shim level? I would have thought that the runwasi library would be the one that supports the OCI runtime spec and all that the shims themselves would have to do is to handle the module that needs to run.

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (rumpl) [reply crates/containerd-shim-wasmedge/src/instance.rs:225]: Yeah, the `DefaultExecutor` is the one that executes the container's process https://github.com/containers/youki/blob/d1e19fdbe3229999c287949e31615b4251c90d17/crates/libcontainer/src/workload/default.rs#L14

**Comment by rumpl** [comment]
I just tested this with docker and I get an error:

```console
docker run --runtime=io.containerd.wasmedge.v1 was
docker: Error response from daemon: failed to create container: unknown.
ERRO[0000] error waiting for container:   
```

In the daemon logs I don't see anything that could help

```console
DEBU[2023-03-28T17:31:39.512641886+02:00] event                                         module=libcontainerd namespace=moby topic=/tasks/create
ERRO[2023-03-28T17:31:39.512738900+02:00] failed to process event                       container=80470a7712ba6ba97f29169ae886d45620569d8fd5a99a52cb7c5ffb391eb078 error="could not find container 80470a7712ba6ba97f29169ae886d45620569d8fd5a99a52cb7c5ffb391eb078: No such container: 80470a7712ba6ba97f29169ae886d45620569d8fd5a99a52cb7c5ffb391eb078" event=create event-info="{80470a7712ba6ba97f29169ae886d45620569d8fd5a99a52cb7c5ffb391eb078 80470a7712ba6ba97f29169ae886d45620569d8fd5a99a52cb7c5ffb391eb078 0 0 0001-01-01 00:00:00 +0000 UTC <nil>}" module=libcontainerd namespace=moby
```

I'll keep digging some more.

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (rumpl) [inline crates/containerd-shim-wasmedge/src/instance.rs:260]: The docs say that this value should be a unique ID:

https://github.com/containerd/runwasi/blob/fff59017205828e1835ea7f699b7a0c5e7b924be/crates/containerd-shim-wasm/src/sandbox/instance.rs#L105-L108

**Comment by rumpl** [comment]
Little update, I don't know yet why this fails. With docker the container creation fails, I'm trying to figure out where and why it fails. Somewhere we get a `ECHILD` error deep inside youki's code.

It doesn't really surprise me though, the container creation sets up all the OCI stuff but we are already setting some of the things in our sandbox...

**Comment by rumpl** [comment]
Right, in this case it's the `waitpid` that fails:

https://github.com/containers/youki/blob/04de7bf3b144da81f991d5756d5b284d4c87cec3/crates/libcontainer/src/process/container_main_process.rs#L110-L119

Removing these lines from youki make it work (for a hello-world example):
```console
$ docker run --runtime=io.containerd.wasmedge.v1 rumpl/was
Hello, world!
```

**Comment by ipuustin** [comment]
> Right, in this case it's the `waitpid` that fails:
> 
> https://github.com/containers/youki/blob/04de7bf3b144da81f991d5756d5b284d4c87cec3/crates/libcontainer/src/process/container_main_process.rs#L110-L119
> 
> Removing these lines from youki make it work (for a hello-world example):

You mean `waitpid()` returns an error? The waitpid docs say:

> [ECHILD]
> The calling process has no existing unwaited-for child processes.

Is this the errno that `waitpid()` sets? Youki could check if the process has already been waited by runwasi, because as far as I understand the waitpid check is for synchronization purposes. (The same check maybe should be added to runwasi too so that there wouldn't be a race condition.)

**Comment by rumpl** [comment]
> Is this the errno that `waitpid()` sets?

Yes, `waitpid` returns the `ECHILD` error in that code with docker

**Comment by utam0k** [comment]
@rumpl hmm... Please give me instructions on how to reproduce this error.

**Comment by rumpl** [comment]
@utam0k You can download the 24.0.0-beta-1 version of dockerd from download.docker.com (for example for Debian it would be here: https://download.docker.com/linux/debian/dists/bullseye/pool/test/amd64/docker-ce-cli_24.0.0~beta.1-1~debian.11~bullseye_amd64.deb, you'll have to look around for your dist or use the static ones).

Then you can:

* Edit the `/etc/docker/daemon.json` file and add `{ "features": { "containerd-snapshotter": true } }`, 
* Run the daemon with `sudo ./dockerd -D -H unix:///tmp/docker.sock --data-root /tmp/dockerd --pidfile /tmp/docker.pid`
* Create a context that will connect to this daemon: `docker context create test --docker "host=unix:///tmp/test.sock"`
* Use that context: `docker context use test` 
* Run a wasm workload with docker.

Once you finished you can go back to your original context: `docker context use default`

**Comment by squillace** [comment]
hey @utam0k I pinged you on discord, but dropping it here, too: we'd love to work with you and @rumpl to get this working on Docker Desktop -- if we can get that, we can merge and move on to showing the beautiful magic. Is that something you can work on? We totally understand if not -- life is complex -- but then we might need to reopen the repo to merge other prs and try this one again later..... 

What do you think?

**Comment by utam0k** [comment]
> hey @utam0k I pinged you on discord, but dropping it here, too: we'd love to work with you and @rumpl to get this working on Docker Desktop -- if we can get that, we can merge and move on to showing the beautiful magic. Is that something you can work on? We totally understand if not -- life is complex -- but then we might need to reopen the repo to merge other prs and try this one again later.....
> 
> What do you think?

I have already replied to that via discord.

**Comment by utam0k** [comment]
> > Is this the errno that `waitpid()` sets?
> 
> Yes, `waitpid` returns the `ECHILD` error in that code with docker

@rumpl How did you know `ECHILD`?

**Comment by utam0k** [comment]
@rumpl cc: @rumpl  
I send out the PR to address that case.
https://github.com/containers/youki/pull/1777

**Comment by utam0k** [comment]
@rumpl PTAL

File: crates/containerd-shim-wasmedge/src/instance.rs
Comment (utam0k) [reply crates/containerd-shim-wasmedge/src/instance.rs]: @rumpl 
I have updated this branch to fix it! Thanks for your review.

File: Cargo.toml
Comment (rumpl) [inline Cargo.toml:34]: Since [this PR](https://github.com/containers/youki/pull/1777) is merged now maybe this can be containers/youki rather than your fork?

**Comment by 0xE282B0** [comment]
Same result here:
> docker: Error response from daemon: failed to load spec: unknown.
> ERRO[0000] error waiting for container:                 

The last thing that is executed is the [ContainerBuilder[...].build()?;](
https://github.com/utam0k/runwasi/blob/libcontainer-wasmedge/crates/containerd-shim-wasmedge/src/instance.rs#L224-L233) when using docker. However, when running the same container with only containerd it is working.

@utam0k, could it be something in the bundle causing that?

**Comment by rumpl** [comment]
Just spent some more time with this, is annoying that anyhow only shows the last error in the chain, the actual error I get is:

```
Apr 06 18:01:45 nuque containerd[528]: [ERROR] ERROR: failed to load spec
Apr 06 18:01:45 nuque containerd[528]: [ERROR] because: failed to validate runtime spec
Apr 06 18:01:45 nuque containerd[528]: [ERROR] because: runtime spec has incompatible version '1.1.0-rc.1'. Only 1.0.X is supported
```

**Comment by 0xE282B0** [comment]
@rumpl, You're right, bug hunting it's a bit cumbersome. Fortunately, the error you're mentioning was the last that blocks the container from running. Now I have at least a simple Wasm container running with moby.

@utam0k, I don't know what implications the version change has for youki but when I [disable the version check](
https://github.com/0xE282B0/youki/commit/b3fe971082569afd4801c824a6d4b6d04b52bebf) it is working for me. Shall we open an issue for youki to support `1.1.x` as well?

**Comment by utam0k** [comment]
@rumpl @0xE282B0 :wave: Thanks for your check.
I will check it this weekend.

**Comment by utam0k** [comment]
@0xE282B0 @rumpl I have put the commit to fix it!

**Comment by 0xE282B0** [comment]
LGTM, I'm able to run a simple Wasm container with Moby v24.0.0-beta.1 on Ubuntu 22.10 x64. 

The log is looking similar to the old containerd-shim-wasmedge version:


Well done! 🥳

I'll test it with different Kubernetes distributions on Tuesday, but when it is working with Docker there should not be a big problem.

**Review by Mossaka — APPROVED** [review]

**Review by cpuguy83 — APPROVED** [review]
LGTM

We can follow up with any necessary fixes.

**Comment by utam0k** [comment]
@cpuguy83 Thanks for the hard PR review 🙇

**Comment by cpuguy83** [comment]
Thanks for putting in all the effort! Great work.

**Pull request merged**
